packages/firestore/src/local/indexeddb_remote_document_cache.ts

```diff
@@ -171,16 +171,22 @@ export class IndexedDbRemoteDocumentCache implements RemoteDocumentCache {
     );
     let firstIteration = true;
 
-    return documentChangesStore(transaction)
+    const changesStore = documentChangesStore(transaction);
+    return changesStore
       .iterate({ range }, (_, documentChange) => {
         if (firstIteration) {
           // If our client was throttled for more than 30 minutes, another
           // client may have garbage collected the remote document changelog.
           if (this._lastProcessedDocumentChangeId + 1 !== documentChange.id) {
-            return PersistencePromise.reject(
-              new FirestoreError(
-                Code.DATA_LOSS,
-                REMOTE_DOCUMENT_CHANGELOG_GARBAGE_COLLECTED_ERR_MSG
+            // Reset the `lastProcessedDocumentChangeId` to allow further
+            // invocations to successfully return the changes after this
+            // rejection.
+            return this.synchronizeLastDocumentChangeId(changesStore).next(() =>
```

Comment:
Just to verify, did you consider just letting this iteration continue so that line 199 can correctly update _lastProcessedDocumentChangeId, and then we could (I think?) get rid of the synchronizeLastDocumentChangeId?   I think the reduced code complexity would outweigh any potential performance saving of getting to skip the changedKeys accounting...